### ISSUE TITLE
Possible bug: posix_data_source_impl::close() may (?) lead to data to be read to a destroyed buffer

### ISSUE THREAD

**Issue by nyh** [scylladb/seastar#779]
This is a *theoretical* bug - I still don't have proof it exists, but I want to document what I know so far. This theory is based on an observed bug in the ScyllaDB project, https://groups.google.com/g/scylladb-dev/c/l7X9AsdNopU/m/AqmWs8JEAgAJ (which I cannot yet reproduce locally). In that observed bug, we have a Seastar HTTP server which checks the request header and decides to close the connection - before reading the entire connection - and crashes.

I begin with two observations about Seastar's (Posix) socket implementation:

1. The `posix_data_source_impl::close() ` method (posix-stack.cc) doesn't wait for anything. It runs shutdown(fd, SHUT_RD) immediately, and returns a ready future.
2. There is no `connected_socket::close()`.

So now consider what Seastar HTTP server code does when it decides to abort a connection before having read all the input from the socket.  `connection::process()` (in httpd.cc) waits for `_read_buf.close()` (which is ends up calling `posix_data_source_impl::close() `) and `_write_buf.close()`, but as mentioned in observations 1 and 2,  we don't wait any more - at that point, the input stream (_read_buf) can be destroyed, and a bit later - also the connected socket may be destroyed.

My theory (which I haven't proved) is that right after the `_read_buf.close()` the read buffer is destroyed but the file descriptor is not yet closed (it was just SHUT_RD). At that point the reactor may read some data already available in the kernel's socket buffer (SHUT_RD doesn't prevent that! See https://stackoverflow.com/questions/740817/behavior-of-shutdownsock-shut-rd-with-tcp) but since the read buffer was destroyed, the theory is that the reactor will read into a destroyed buffer, and crash. 

What I don't understand, though (and makes my whole theory sound a bit dubious), is why would the reactor even try to read data from the socket. Do we have read-ahead on network sockets? If not, the reactor shouldn't read from it (and it shouldn't even have POLLIN enabled) unless there is a pending read issued by the HTTP server - and there is none.

**Comment by avikivity** [comment]
We don't have network read-ahead.

The problem can happen if data_source::read() and data_source::close() are running concurrently (which they never should).

**Comment by nyh** [comment]
> We don't have network read-ahead. 
> The problem can happen if data_source::read() and data_source::close() are running concurrently (which they never should).

I can't find any evidence for this in the Seastar http code.... httpd.cc's `connection::read()` calls `read_one()` repeatedly which reads from the socket's input buffer - until _done. Only then, in a  finally(), it calls `_read_buf.close()`. Which seems fine.

Maybe this isn't the problem at all - but I don't have any other theory on that observed Scylla bug that causes a crash in that point in the HTTP server.

**Comment by nyh** [comment]
I was barking up the wrong tree... There appears to be nothing wrong with the network code (Avi, sorry for leading you on this path).
I just realised the Scylla bug report involved an older version of Scylla (and Seastar), and the Seastar there is missing the commit c8c209da30190aaf63f6055801a35dcda37c05e3 by @espindola. This fixed a bug where to generate the error message, we used an already-moved value.
So closing this Seastar issue - it was already solved.
